cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -122,22 +128,24 @@ void put(String key, Value value, Scope scope) {
     Preconditions.checkNotNull(value.toString(), String.format("Null value provided for the key '%s'.", key));
     Preconditions.checkState(nodeName != null, "nodeName cannot be null.");
 
-    List<NodeValue> nodeValueList = tokenValueMap.get(scope).get(key);
-    if (nodeValueList == null) {
-      nodeValueList = Lists.newArrayList();
-      tokenValueMap.get(scope).put(key, nodeValueList);
-    }
+    synchronized (sharedLock) {
+      List<NodeValue> nodeValueList = tokenValueMap.get(scope).get(key);
```

Comment:
generally we don't include types in variable names right? nodeValues would be a better name.